File: go/libraries/shared/server/mainproc.go
package server

import (
	"log"
	"net/http"
	"os"
	"os/signal"
)

// MainProc is a main routine for rpc server processes
func MainProc(category string, typex string, registry *Registry, appHandler map[string]http.Handler) {

	// create server
	server, err := registry.NewServer(category, typex, appHandler)
	if err != nil {
		log.Printf("could not start rpc server [%v]", err)
		return
	}

	log.Printf("using address: http://localhost:%d/", server.Address.Port)

	// install stop signal channel
	stopChan := make(chan os.Signal, 2)
	signal.Notify(stopChan, os.Interrupt, os.Kill)

	// wait for stop signal or stopped event of server
	select {
	case <-stopChan:
		log.Printf("stop signal received...")
		break

	case err = <-server.StoppedEvent:
		if err != nil {
			log.Printf("rpc server stopped unexpected [%v]", err)
		}
	}

	// shutdown rpc server gracefully
	if err = server.Unpublish(); err != nil {
		log.Printf("could not shutdown rpc server gracefully [%v]", err)
	}

	// we are done now!
	log.Printf("exiting now... bye!")
}

File: javascript/libraries/shared/src/server/mainproc.ts
// NODE ONLY!

import net from 'net';
import process from 'process';
import { Registry } from './registry';
import { newServer, shutdownServer } from './server';

export async function mainProc(
  category: string, type: string, registry: Registry,
  onStarted: ((address: net.AddressInfo) => Promise<void>) | null,
  onShutdown: (() => Promise<void>) | null
): Promise<void> {

  // resume stdin if paused
  const stdinWasPaused = process.stdin.isPaused();
  if (stdinWasPaused) {
    process.stdin.resume();
  }

  try {
    // create server
    const server = await newServer(registry, category, type);

    if (onStarted) {
      await onStarted(server.address);
    }

    // wait for stop signal
    await new Promise(resolve => {
      function onSigInt() {
        process.off('SIGINT', onSigInt);
        process.off('SIGTERM', onSigInt);
        resolve();
      }
      process.on('SIGINT', onSigInt);
      process.on('SIGTERM', onSigInt);
    });

    // shutdown rpc server gracefully
    await shutdownServer(server);

    if (onShutdown) {
      await onShutdown();
    }

    // we are done now!
  }
  finally {

    // reset stdin state
    if (stdinWasPaused) {
      process.stdin.pause();
    }
  }
}
